packages/Search/src/DTK_LinearBVH_fwd.hpp

```diff
@@ -0,0 +1,139 @@
+
+namespace DataTransferKit
+{
+
+template <typename DeviceType>
+class BoundingVolumeHierarchy
+{
+  public:
+    using device_type = DeviceType;
+    using bounding_volume_type = Box;
+    using size_type = typename Kokkos::View<Node *, DeviceType>::size_type;
+
+    BoundingVolumeHierarchy() = default; // build an empty tree
+
+    template <typename Primitives>
+    BoundingVolumeHierarchy( Primitives const &primitives );
+
+    KOKKOS_INLINE_FUNCTION
+    size_type size() const { return _leaf_nodes.extent( 0 ); }
+
+    KOKKOS_INLINE_FUNCTION
+    bool empty() const { return size() == 0; }
+
+    KOKKOS_INLINE_FUNCTION
+    bounding_volume_type bounds() const
+    {
+        // NOTE should default constructor initialize to an invalid geometry?
+        if ( empty() )
+            return bounding_volume_type();
+        return getBoundingVolume( this->getRoot() );
+    }
+
+    template <typename Predicates, typename... Args>
+    inline void query( Predicates const &predicates, Args &&... args ) const
+    {
+        // FIXME lame placeholder for concept check
+        static_assert( Kokkos::is_view<Predicates>::value, "must pass a view" );
+        using Tag = typename Predicates::value_type::Tag;
+        Details::BoundingVolumeHierarchyImpl<DeviceType>::queryDispatch(
+            Tag{}, *this, predicates, std::forward<Args>( args )... );
+    }
+
+    // private:
+  public:
+    Kokkos::View<Node *, DeviceType, Kokkos::MemoryUnmanaged> getLeafNodes()
+    {
+        return _leaf_nodes;
+    }
+
+    /**
+     * Return the root node of the BVH.
+     */
+    KOKKOS_INLINE_FUNCTION Node *getRoot()
+    {
+        if ( empty() )
+            return nullptr;
+        return ( size() > 1 ? _internal_nodes : _leaf_nodes ).data();
+    }
+    KOKKOS_INLINE_FUNCTION Node const *getRoot() const
+    {
+        if ( empty() )
+            return nullptr;
+        return ( size() > 1 ? _internal_nodes : _leaf_nodes ).data();
+    }
+    /**
+     * Return the bounding volume of a node of the BVH.
+     */
+    KOKKOS_INLINE_FUNCTION Box &getBoundingVolume( Node *node )
+    {
+        return node->bounding_box;
+    }
+    KOKKOS_INLINE_FUNCTION Box const &
+    getBoundingVolume( Node const *node ) const
+    {
+        return node->bounding_box;
+    }
+    /**
+     * Return true if the node is a leaf.
+     */
+    KOKKOS_INLINE_FUNCTION bool isLeaf( Node const *node ) const
+    {
+        return ( node->children.first == nullptr );
+    }
+    /**
+     * Return the index of a leaf node.
+     */
+    // FIXME: debate on whether we only need getPermutationIndex for leaf
+    // nodes, and what to do (if we need to) for internal
+    KOKKOS_INLINE_FUNCTION int getPermutationIndex( Node const *leaf ) const
+    {
+        static_assert( sizeof( size_t ) == sizeof( Node * ),
+                       "Conversion is a bad idea if these sizes do not match" );
+        return reinterpret_cast<size_t>( leaf->children.second );
```

Comment:
Update the comment above and make it clear that the return is garbage if you pass an internal node.  As a matter of fact, I'd rather rename it to something like `getLeafPermutationIndex()`

Reply:
Renamed to `getLeafPermutationIndex()`. Still uncomfortable that we don't have a runtime check here. We should probably add
```c++
DTK_REQUIRE(leaf->children.first == nullptr);
```
Also, not 100% sure this function belongs to BVH, and not to TreeTraversal.